nn/conformer.py

```diff
@@ -0,0 +1,215 @@
+"""
+Conformer code.
+Ref: https://arxiv.org/abs/2005.08100
+"""
+
+from typing import Tuple, List, Union, Callable, Optional
+from .. import nn
+
+
+class _PositionwiseFeedForward(nn.Module):
+  """
+  Conformer position-wise feedforward neural network layer
+      FF -> Activation -> Dropout -> FF
+  """
+
+  def __init__(self, out_dim: int, dim_ff: int, dropout: float, activation: Callable[[nn.LayerRef], nn.LayerRef]):
+    """
+    :param out_dim:
+    :param dim_ff:
+    :param dropout:
+    :param activation:
+    """
+    super().__init__()
+
+    self.dropout = dropout
+    self.activation = activation
+
+    self.linear_ff = nn.Linear(n_out=dim_ff)
+    self.linear_out = nn.Linear(n_out=out_dim)
+
+  def forward(self, inp: nn.LayerRef) -> nn.LayerRef:
+    x_ff1 = self.linear_ff(inp)
+    x_act = self.activation(x_ff1)
+    x_drop = nn.dropout(x_act, dropout=self.dropout)
+    x_ff2 = self.linear_out(x_drop)
+    return x_ff2
+
+
+class _ConformerConvBlock(nn.Module):
+  """
+  Conformer convolution block
+      FF -> GLU -> depthwise conv -> BN -> Swish -> FF
+  """
+
+  def __init__(
+        self, out_dim: int, kernel_size: int, batch_norm_eps: float = 1e-5, batch_norm_momentum: float = 0.1,
+        batch_norm_other_opts=None):
```

Comment:
See my other comments on batch norm options.

Reply:
ok but the values for `batch_norm_eps` and `batch_norm_momentum` won't be specified in `nn.BatchNorm` right? otherwise we can not pass them in `batch_norm_opts` as this would be duplicate params.

Reply:
What do you mean? They are specified in `nn.BatchNorm`, with reasonable defaults. Where are they duplicated?

Reply:
Yeah nevermind. Should be resolved now.